from limix._display import session_line

from .._data import conform_dataset, asarray as _asarray
from .._display import session_block
from ._assert import assert_finite


def sscan(G, y, E, M=None, idx=None, tests=None, verbose=True):
    r"""
    Structured linear mixed model that accounts for genotype-environment interactions.

    StructLMM [MC18]_ extends the conventional linear mixed model by including an
    additional per-individual effect term that accounts for genotype-environment
    interaction, which can be represented as an nÃ—1 vector, ð›ƒâ‚.
    The model can be cast as

    .. math::

        ð² = ð™¼ð›‚ + ð ð›½â‚€ + ð âŠ™ð›ƒâ‚ + ðž + ð›†, ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\\
        \text{where}~~ ð›ƒâ‚âˆ¼ð“(ðŸŽ, ð“‹â‚€Î£),~~ ðžâˆ¼ð“(ðŸŽ, ð“‹â‚Î£),~~\text{and}~~ ð›†âˆ¼ð“(ðŸŽ, ð“‹â‚‚ð™¸),

    where ð›½â‚€ denotes the effect size of a conventional persistent genetic effect
    component and ð›ƒâ‚ is a vector of per-individual effect sizes to account for
    heterogeneous genetic effects. The same environmental covariance Î£ is used to
    account for genetic-environment interaction and for additive environmental effects.
    Vector ð  is the variant of interest to test for association and/or interaction.
    The term ð ð›½â‚€ represents the genetic effect, while term ð âŠ™ð›ƒâ‚ represents the
    genetic-environment interaction effect.
    The parameters of the model are ð›½â‚€, ð›ƒâ‚, ð“‹â‚€, ð“‹â‚, and ð“‹â‚‚.

    It performs score tests for association and interaction, respectively::

    - Hâ‚€ vs Hâ‚: testing for ð›½â‚€ â‰  0 while ð›ƒâ‚ = ðŸŽ
    - Hâ‚€ vs Hâ‚‚: testing for [ð›½â‚€ ð›ƒâ‚] â‰  ðŸŽ

    Parameters
    ----------
    G : nÃ—m array_like
        Genetic candidates.
    Y : nÃ—p array_like
        Rows are samples and columns are phenotypes.
    E : nÃ—ðš” array_like
        Samples-by-environments design matrix.
    M : nÃ—c array_like, optional
        Covariates matrix.
    idx : list, optional
        List of candidate indices that defines the set of candidates to be used in the
        tests.
        It defaults to ``range(n)`` such that every column of matrix G
    tests : list, optional
        List of tests to be performed.
        The possible values are ``"inter"`` and ``"assoc"``.
        It defaults to ``["inter"]``.
    verbose : bool, optional
        ``True`` to display progress and summary; ``False`` otherwise.

    References
    ----------
    .. [MC18] Moore, R., Casale, F. P., Bonder, M. J., Horta, D., Franke, L., Barroso,
        I., & Stegle, O. (2018). A linear mixed-model approach to study multivariate
        geneâ€“environment interactions (p. 1). Nature Publishing Group.
    """
    from struct_lmm import StructLMM
    from numpy import asarray, zeros, hstack

    if tests is None:
        tests = set(["inter"])
    else:
        tests = set(tests)

    remain = tests - set(["inter", "assoc"])
    if len(remain) > 0:
        raise ValueError(f"Unrecognized test parameters: {remain}.")

    with session_block("QTL analysis", disable=not verbose):

        with session_line("Normalising input... ", disable=not verbose):

            data = conform_dataset(y, M, G=G)

        y = data["y"]
        M = data["M"]
        G = data["G"]
        K = data["K"]

        assert_finite(y, M, K)

        E = _asarray(E, "env", ["sample", "env"])

        if "assoc" in tests:
            slmm = StructLMM(y, M, E)
            slmm.fit(verbose=False)

        _pvi = zeros(G.shape[1])
        _pva = zeros(G.shape[1])
        for snp in range(G.shape[1]):
            x = asarray(G[:, [snp]], float)

            if "inter" in tests:
                M1 = hstack((M, x))
                slmi = StructLMM(y, M1, E)
                slmi.fit(verbose=False)
                _pvi[snp] = slmi.score_2dof_inter(x)

            if "assoc" in tests:
                _pva[snp] = slmm.score_2dof_assoc(x)

    return _pvi, _pva
